### PR TITLE
Improve error reporting and disallow f(**kwargs, *args)

### PR DESCRIPTION
**Pull request by pablogsal**
Closes: #45 
relates-to: #43

### CODE REVIEW

File: Parser/pegen/pegen.c [hunk 1]
```diff
@@ -720,6 +720,26 @@ seq_insert_in_front(Parser *p, void *a, asdl_seq *seq)
     return new_seq;
 }
 
+/* Creates a copy of seq and prepends a to it */
+asdl_seq *
+concatenate_sequences(Parser *p, asdl_seq *a, asdl_seq *b)
```
Comment (lysnikolaou) [inline Parser/pegen/pegen.c:725]: There's already the private function `_join_seqs`, which you can use.

File: Grammar/python.gram [hunk 2]
```diff
@@ -447,13 +454,19 @@ args[expr_ty]:
                      : CHECK(singleton_seq(p, a)),
                  (b) ? ((expr_ty) b)->v.Call.keywords : NULL,
                  EXTRA) }
-kwargs[asdl_seq*]: a=','.kwarg+ { a }
+kwargs[asdl_seq*]:
+    | a=','.kwarg+ & ')' { a }
```
Comment (lysnikolaou) [inline Grammar/python.gram:458]: Is this lookahead needed?

File: Grammar/python.gram [hunk 3]
```diff
@@ -447,13 +454,19 @@ args[expr_ty]:
...
+    | b=','.kwarg_no_starred+ { b }
+    | a=','.kwarg+ ',' b=','.kwarg_no_starred+ { CHECK(concatenate_sequences(p, a, b)) }
```
Comment (lysnikolaou) [inline Grammar/python.gram:460]: No need for a top-level `CHECK` call.
```suggestion
    | a=','.kwarg+ ',' b=','.kwarg_no_starred+ { concatenate_sequences(p, a, b) }
```

File: Grammar/python.gram (hunk 2 above)
Comment (pablogsal) [reply Grammar/python.gram:458]: As far as I understand yes because if not there it can consume partially the argument sequence and won't try the next option

File: Grammar/python.gram [hunk 4]
```diff
@@ -429,7 +429,14 @@ yield_expr[expr_ty]:
     | 'yield' a=[star_expressions] { _Py_Yield(a, EXTRA) }
 
 arguments[expr_ty] (memo):
-    | a=args [','] { a }
+    | a=args [','] &')' { a }
+    | incorrect_arguments [',']
+
+incorrect_arguments:
+    | args ',' '*' { raise_syntax_error(p, "iterable argument unpacking follows keyword argument unpacking") }
```
Comment (lysnikolaou) [inline Grammar/python.gram:436]: This would probably disallow things like `f(x=1, *[], **{})`. A negatve lookahead for a star thereafter should do it though.

Also, I feel somewhat awkward to have `raise_syntax_error` return an `int` and then assign it to a variable of type `void *`.

File: Grammar/python.gram (hunk 2 above)
Comment (pablogsal) [reply Grammar/python.gram:463]: For instance, try `f(*args, x=3, **kwargs)` without that lookahead

File: Grammar/python.gram (hunk 4 above)
Comment (pablogsal) [reply Grammar/python.gram:441]: > This would probably disallow things like `f(x=1, *[], **{})`. A negatve lookahead for a star thereafter should do it though.

Yes, but the idea is that that alternative must have parsed previously by other rule with higher precedence so you could never reach that option. I was missing a lookahead to make this work (added in 11163f4):

```
❯ PYTHONPARSER=new ./python 
Python 3.9.0a5+ (heads/some_errors-dirty:312f116214, Apr  5 2020, 23:23:07) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> def f(*a, **k): ...
... 
>>> f(x=1, *[], **{})
```

Comment (pablogsal) [reply Grammar/python.gram:441]: Whops, but there is still something wrong with assignment expressions as my latest commit didn't makes now `spam(a=1, (b := 2)` fail :(

File: Grammar/python.gram [hunk 5]
```diff
@@ -263,8 +263,13 @@ expression[expr_ty] (memo):
...
     | lambdef
 
+expression_for_lambda[expr_ty] (memo):
+    | a=disjunction 'if' b=disjunction 'else' c=expression { _Py_IfExp(b, a, c, EXTRA) }
+    | disjunction
+    | lambdef
+
 lambdef[expr_ty]:
-    | 'lambda' a=[lambda_parameters] ':' b=expression { _Py_Lambda((a) ? a : CHECK(empty_arguments(p)), b, EXTRA) }
+    | 'lambda' a=[lambda_parameters] ':' b=expression_for_lambda { _Py_Lambda((a) ? a : CHECK(empty_arguments(p)), b, EXTRA) }
```
Comment (lysnikolaou) [inline Grammar/python.gram:272]: Why is this needed?

Comment (pablogsal) [reply Grammar/python.gram:272]: Sorry, that was me working on a different issue and forgot to stash. The issue is:

```
>>> (lambda: x:=2)
  File "<stdin>", line 1
SyntaxError: cannot use assignment expressions with lambda
>>> 
```

File: Grammar/python.gram (hunk 4 above)
Comment (lysnikolaou) [reply Grammar/python.gram:436]: Shouldn't that fail though?

Comment (pablogsal) [reply Grammar/python.gram:436]: Yeah, but it fails now with a different error

Comment (lysnikolaou) [reply Grammar/python.gram:436]: Oh, okay.

File: Parser/pegen/pegen.c (hunk 1 above)
Comment (lysnikolaou) [reply Parser/pegen/pegen.c:727]: Do we need this? Couldn't we just make either one of these public and just use that one?

Comment (pablogsal) [reply Parser/pegen/pegen.c:727]: Good point. Sorry, I had my mind on the error messages and fixed this too quickly :S

Comment (lysnikolaou) [reply Parser/pegen/pegen.c:727]: No worries!

File: Grammar/python.gram (hunk 4 above)
Comment (pablogsal) [reply Grammar/python.gram:436]: I think I managed to fix it, but I am a bit tired and I may have missed something

Comment (lysnikolaou) [reply Grammar/python.gram:436]: One thing I found that's failing is `f(1, x=2, 3)`. The text is `SyntaxError: positional argument follows keyword argument unpacking`. I'm also having difficulties coming up with something that's universal.

Comment (pablogsal) [reply Grammar/python.gram:436]: :( I think this may be indicating that we are missing some other shape for this piece of the grammar. Maybe we would need to insert the errors among the normal rules instead of deferring them to the end.....

Comment (lysnikolaou) [reply Grammar/python.gram:436]: That's what I thought as well. I could work on that tomorrow. 2:30 is too late for this. :laughing: 

On another note, would it be okay to make `raise_syntax_error` return NULL, or is there a function somewhere that requires that it returns an `int`?

Comment (pablogsal) [reply Grammar/python.gram:436]: > That's what I thought as well. I could work on that tomorrow. 2:30 is too late for this. laughing

Yeah, the beginning of this week is going to be a bit challenging for me so if you are able to keep iterating that would be great! It would be good to see if @gvanrossum has some insight on this as it seems that maybe some other refactoring of the rule could help here.

> On another note, would it be okay to make `raise_syntax_error` return NULL, or is there a function somewhere that requires that it returns an `int`?

Yeah, that's something to fix at the end. But there are multiple ways to fix that.

Comment (lysnikolaou) [reply Grammar/python.gram:436]: > Yeah, the beginning of this week is going to be a bit challenging for me so if you are able to keep iterating that would be great! It would be good to see if @gvanrossum has some insight on this as it seems that maybe some other refactoring of the rule could help here.

I'll have plenty of time (not only this week, but probably for the next three weeks or so), so I'll certainly try my best.

> Yeah, that's something to fix at the end. But there are multiple ways to fix that.
 
:+1:

**Comment by lysnikolaou** [comment]
I just pushed a version that should work correctly, which uses a helper function. I realize that's a bit ugly, so I would like some feedback. If you think we can go ahead with his, I will add some tests tomorrow.

**Comment by pablogsal** [comment]
> I just pushed a version that should work correctly, which uses a helper function. I realize that's a bit ugly, so I would like some feedback. If you think we can go ahead with his, I will add some tests tomorrow.

I can try to review this tomorrow but in the meanwhile check out this comment for a potential alternative way of handling errors:

https://github.com/we-like-parsers/cpython/issues/43#issuecomment-610116884

File: Parser/pegen/pegen.c [hunk 6]
```diff
@@ -1118,8 +1137,8 @@ star_etc(Parser *p, arg_ty vararg, asdl_seq *kwonlyargs, arg_ty kwarg)
     return a;
 }
 
-static asdl_seq *
-_join_seqs(Parser *p, asdl_seq *a, asdl_seq *b)
+asdl_seq *
+join_sequences(Parser *p, asdl_seq *a, asdl_seq *b)
```
Comment (gvanrossum) [inline Parser/pegen/pegen.c:1141]: This reminds me that we have a lot of non-static functions whose name doesn't start with `_Py`. Shouldn't we rename those? I believe it's still a strict rule to use `_Py` for public functions that aren't part of the public API.

Maybe this is something we'll need to clean up later?

Comment (pablogsal) [reply Parser/pegen/pegen.c:1141]: Yeah, very likely we need some renaming and maybe to use some private API macros. As long as `make smelly` does not complain, we should be good.

Comment (lysnikolaou) [reply Parser/pegen/pegen.c:1141]: It does, 64 times!

Comment (pablogsal) [reply Parser/pegen/pegen.c:1141]: 😅

Comment (pablogsal) [reply Parser/pegen/pegen.c:1141]: And will suggest leaving it for the end as this is likely just some renaming + surrounding with API macros.

**Comment by gvanrossum** [comment]
To resolve the conflict, `make regen-pegen`. (Maybe that rule should be given proper dependencies in Makefile.pre.in now?)
